Feat/docs 5469 passwordless db connect (#1157)
auth0/docs-v2: main/docs/authenticate/database-connections/passwordless-authentication-for-db-connect.mdx

---
title: Passwordless Authentication on Database Connections
sidebarTitle: Passwordless Authentication
description: Learn how to configure true passwordless authentication on a new or existing database connection.
---
import { ReleaseStageNotice } from "/snippets/ReleaseStageNotice.jsx"

<ReleaseStageNotice
  feature="Passwordless Authentication on Database Connections"
  stage="ea"
  contact="support"
  terms="true"
/>

> Auth0 allows you to configure email and phone-based passwordless authentication directly on a [database connection](/docs/authenticate/database-connections). Rather than create a separate connection for one-time password (OTP) authentication, your ability to offer passwordless authentication to users in database connections makes it easier to deliver flexible login and signup to end-users while reducing implementation complexity. Passwordless for database connections is not supported for use with Classic Login.

Reviewer: ```suggestion
Auth0 allows you to configure email and phone-based passwordless authentication directly on a [database connection](/docs/authenticate/database-connections). Rather than creating a separate connection for one-time password (OTP) authentication, you can offer passwordless login directly from your database connection. This reduces implementation complexity and simplifies the login experience for end-users. Passwordless for database connections is not supported for use with Classic Login.
```
this sentence a bit overloaded. chopped it and tightened

Reviewer: The last sentence:
>Passwordless for database connections is not supported for use with Classic Login.

seemes orphaned a bit. maybe we should move into a callout?

## How it works

> Auth0 uses an [Identifier-First Authentication](https://auth0.com/docs/authenticate/login/auth0-universal-login/identifier-first) approach in which users are verified with OTP authentication:

Reviewer: ```suggestion
Auth0 uses an [Identifier-First Authentication](/docs/authenticate/login/auth0-universal-login/identifier-first) approach in which users are verified with OTP authentication:
```

1. An end user enters an identifier to login with [Auth0's Universal Login](/docs/authenticate/login/auth0-universal-login).
2. Auth0 looks up the authentication methods for each identifier configured on your database connection.

> 3. The user is presented with the most suitable option based on the `default_method` and available authentication methods (e.g., receive an OTP by email or phone or provide a password).

Reviewer: >The user is presented with the most suitable option based on the `default_method`

does that just mean the user is presented with whatever is set to default_method?

> 3. The user is presented with the most suitable option based on the `default_method` and available authentication methods (e.g., receive an OTP by email or phone or provide a password).

Reviewer: ```suggestion
3. The user is presented with the most suitable option based on the `default_method` and available authentication methods (for example, receive an OTP by email or phone or provide a password).
```
removing latin abbreviation

<Card title="Before you start">

* Configure your tenant to use the [Identifier-First Authentication Profile](https://auth0.com/docs/authenticate/login/auth0-universal-login/identifier-first).
* If you plan to use phone-based OTP (SMS or Voice), you must enable the [Unified Phone Experience](/docs/customize/phone-messages/unified-phone/configure-unified-phone) setting under [**Branding > Phone Provider**](https://manage.auth0.com/#/phone/templates/phone/provider).
* [Attributes](/docs/authenticate/database-connections/activate-and-configure-attributes-for-flexible-identifiers) must be enabled on the database connection.
* If you plan to use existing (legacy) passwordless connections in conjunction with passwordless authentication on database connections, and you don't want to configure your existing passwordless connections to use the Unified Phone Provider:
  1. Navigate to [**Auth0 Dashboard > Authentication > Passwordless**](https://manage.auth0.com/#/connections/passwordless).
  2. Select **Configure** by SMS.
  3. Make sure the **Use Tenant-Level Messaging Provider** setting is disabled.

</Card>

## Configure attributes (identifiers)

> When configuring email and phone-based passwordless authentication on database connections, you must first determine which attribute(s) you want end-users to provide as during signup and login.

Reviewer: ```suggestion
When configuring email and phone-based passwordless authentication on database connections, you must first determine which attribute(s) you want end-users to provide during signup and login.
```

2. Enter a unique name for your connection.
3. Choose one or more attributes for end-users to log in or sign up.
4. Select your authentication method(s). You can further configure these methods once you've created the connection.
5. Toggle on **Disable Sign Ups** if you don't want users to sign up using public endpoints. 
6. Toggle on **Promote Connection to Domain Level** if you want to use this connection with third-party applications.
7. Select **Create**.

For a true passwordless connection, complete the following additional steps:

1. In your new connection, select the **Attributes** tab.
2. To disable Username as an identifier, select **Configure** and toggle off **Use Username as Identifier**.
3. To configure Email and Phone identifiers, select **Configure**. 
  * For Email attributes, select **One-Time Password (OTP)** under Verification Methods and enable **Verify email on sign up** for true passwordless configuration. This ensures `email_verified` is automatically set so users are always prompted for OTP on login and signup.
4. Select **Save**.

> 5. Configure settings authentication methods that correspond to the chosen identifiers. You cannot disable password unless you have `phone_otp` and/or `email_otp` configured.

Reviewer: Missing or extra word here

Author: In the next commit

6. Under Password settings, select **Policy** and choose **Block** for:

   * Password on Login
   * Password on Signup
   * Self-service change password (updates automatically)

7. Toggle on **Support users without a password**.
   <Warning>
   You may get an error if you do not toggle on **Support users without a password**.
   </Warning>
8. Select **Save**. In the prompt, select **Continue** to confirm you understand that existing users may be affected.
9. Navigate to the **Applications** tab and enable the connection for your application or API.

</Tab>
<Tab title="Management API" icon="code">

<Callout icon="file-lines" color="#0EA5E9" iconType="regular">

>   To use the Management API, you need a [Management API access token](/docs/secure/tokens/access-tokens/get-management-api-access-tokens-for-production).

Reviewer: ```suggestion
  To use the Management API, you need a [Management API access token](/docs/secure/tokens/access-tokens/management-api-access-tokens/get-management-api-access-tokens-for-production).
```

        }
      },
      "username": {
        "identifier": {
          "active": true
        },
        "profile_required": true,
        "signup": {
          "status": "required"
        },
        "validation": {
          "min_length": 1,
          "max_length": 15,
          "allowed_types": {
            "email": false,
            "phone_number": false
          }
        }
      }
    },
    "authentication_methods": {
      "password": {
        "enabled": true
      },
      "passkey": {
        "enabled": false
      },
      "email_otp": {
        "enabled": true
      },
      "phone_otp": {
        "enabled": true
      }
    }
  }
}'
```

</Tab>
</Tabs>

## Update existing connections

If you have current database connections, update the settings for passwordless in the Auth0 Dashboard or Management API.

<Tabs>
<Tab title="Auth0 Dashboard" icon="browser">

1. **Open the connection**: Navigate to [**Auth0 Dashboard > Authentication > Database**](https://manage.auth0.com/#/connections/database) and select the connection you want to update.
2. **Activate Attributes**: Under the **Attributes** tab, select **Activate** to enable the New Attributes Configuration.

> 3. **Add Email and Phone attributes**: Choose **+ Add Attributes**. If Email is not visible, add Email and Phone Number.

Reviewer: >If Email is not visible, add Email and Phone Number

this seems kind of unclear. maybe, "Choose + Add Attributes and add Email and Phone Number if they are not already present."

   <Callout icon="file-lines" color="#0EA5E9" iconType="regular">
     For a true passwordless (OTP) connection, the username identifier is not supported.
   </Callout>
4. **Enable OTP authentication methods**: Under the **Authentication Methods** tab:
   * Configure Phone to **Allow** Phone OTP and save your changes.
   * Configure Email to **Allow** Email OTP and save your changes.
5. **Block password authentication**: Under Password settings, choose **Policy** and select **Block** for:

>    * Password for login

Reviewer: Above, these are stylized as:

   * Password on Login
   * Password on Signup
   * Self-service change password (updates automatically)

just keep casing consistent in both

-H "Content-Type: application/json" \
-H "Authorization: Bearer YOUR_MANAGEMENT_API_TOKEN" \
-d '{
  "options": {
    "passwordPolicy": "none",
    "authentication_methods": {
      "password": { "enabled": true },
      "passkey": { "enabled": true },
      "email_otp": { "enabled": true },
      "phone_otp": { "enabled": true }
    },
    "attributes": {
      "email": {
        "signup": {
          "status": "required",
          "verification": { "active": true }
        },
        "unique": true,
        "identifier": {
          "active": true,
          "default_method": "email_otp"
        },
        "profile_required": true,
        "verification_method": "otp"
      },
      "phone_number": {
        "signup": {
          "status": "required",
          "verification": { "active": true }
        },
        "identifier": {
          "active": true,
          "default_method": "phone_otp"
        },
        "profile_required": true
      }
    }
  }
}'
```

</Tab>
</Tabs>

## Use Auth0 Actions

For more insight into passwordless authentication factors on database connections, configure attributes with Auth0 Actions.

### Post-login trigger

> The `post-login` trigger fires after a user authenticates but before the authorization server returns a token. Configure `event.authentication` in [Auth0 Dashboard > Actions > Triggers > Post Login](https://manage.auth0.com/#/actions/flows) for the following attributes:

Reviewer: event.authentication is read-only, so you don't really configure it. you can say, "the event.authentication object in ... exposes the following methods:"

</Callout>

Use `event.authentication` to:

* Detect which passwordless factor the user completed (`email`, `sms`, `tel`, or `pwd`)
* Add custom claims to tokens based on the authentication method
* Conditionally run logic based on how the user authenticated

#### Example

The following example reads `event.authentication.methods` to detect which passwordless factor the user completed and adds it as a custom claim on the ID token.

```javascript
exports.onExecutePostLogin = async (event, api) => {
  if (!event.authentication?.methods || event.authentication.methods.length === 0) {
    return;
  }

  if (event.connection.name !== 'YOUR_AUTH0_CONNECTION') {
    return;
  }

  const firstFactor = event.authentication.methods[0];

  if (firstFactor.name === 'email') {
    api.idToken.setCustomClaim('https://your-app.com/auth_method', 'email_otp');
  } else if (firstFactor.name === 'sms' || firstFactor.name === 'tel') {
    api.idToken.setCustomClaim('https://your-app.com/auth_method', 'phone_otp');
  } else if (firstFactor.name === 'pwd') {
    api.idToken.setCustomClaim('https://your-app.com/auth_method', 'password');
  }
};
```

### Post-challenge trigger

> The `post-challenge` trigger fires after users complete a challenge, such as password reset, phone validation, or MFA. Configure `event.authentication` in [Auth0 Dashboard > Actions > Triggers > password-reset-post-challenge](https://manage.auth0.com/#/actions/flows) for the following attributes:

Reviewer: same as above on configuring event.authentication. it's minor, but if you do decide to change above, you can do it here too.

| Attribute | Parameter | Description |
| --- | --- | --- |
| Email | `email` | Password reset with email OTP or magic link. |

> | Phone | `phone_number` | Password reset with phone OTP. |

Reviewer: should probably add password reset to this table since you mention it in the intro

---
title: Passwordless Authentication on Database Connections
sidebarTitle: Passwordless Authentication
description: Learn how to configure true passwordless authentication on a new or existing database connection.
---
import { ReleaseStageNotice } from "/snippets/ReleaseStageNotice.jsx"

<ReleaseStageNotice
  feature="Passwordless Authentication on Database Connections"
  stage="ea"
  contact="support"
  terms="true"
/>

Auth0 allows you to configure email and phone-based passwordless authentication directly on a [database connection](/docs/authenticate/database-connections). Rather than create a separate connection for one-time password (OTP) authentication, your ability to offer passwordless authentication to users in database connections makes it easier to deliver flexible login and signup to end-users while reducing implementation complexity. Passwordless for database connections is not supported for use with Classic Login.

## How it works

Auth0 uses an [Identifier-First Authentication](https://auth0.com/docs/authenticate/login/auth0-universal-login/identifier-first) approach in which users are verified with OTP authentication:

> 1. An end user enters an identifier upon sign-up with [Universal Login](/docs/authenticate/login/auth0-universal-login).

Reviewer 2: Since we're talking about login/authentication, referring to "sign-up" here seems confusing. Perhaps

```suggestion
1. An end user enters an identifier upon login with [Universal Login](/docs/authenticate/login/auth0-universal-login).
```
or something along those lines?

Author: Added to the next commit

Author: Added to next commit

2. Auth0 looks up the identifier against the database connection and determines which authentication methods are available for the user.

> 3. The user is presented with the available options (e.g., receive an OTP by email or phone, provide a username and password, or use a passkey with another authentication method).

Reviewer 2: This wording appears a bit confusing to me. It almost sounds like we look up the user and then do a determination on what the user has previously enrolled with. #3 isn't quite right, we don't present multiple options the way we may do for MFA factors, we just pick one and let the user switch if they want. 
Perhaps we can consider changing to something like:

```suggestion
2. Auth0 looks up the database connection's configured authentication methods to determine which are available for that identifier type.
3. The user is presented with the most suitable option based on the `default_method` and all available auth methods' configuration (e.g., receive an OTP by email or phone or provide a password).
```

Author: Added to next commit

> 4. Auth0 sends a code to the user's email or phone.

Reviewer 2: ```suggestion
4. If OTP was the determined authentication method, Auth0 sends a code to the user's email or phone.
```

> 5. The user enters the code and is authenticated unless passkey is used in addition to OTP.

Reviewer 2: ```suggestion
5. The user enters the code and is authenticated. If applicable, the user is prompted for progressive passkey enrollment.
```

Author: Added to next commit

<Tabs>
<Tab title="Auth0 Dashboard" icon="browser">

1. Navigate to [**Auth0 Dashboard > Authentication > Database**](https://manage.auth0.com/#/connections/database). Choose **Create DB Connection** to create the connection.
2. Enter a unique name for your connection.
3. Choose one or more attributes for end-users to log in or sign up.
4. Select your authentication method(s). You can further configure these methods once you've created the connection.
5. Toggle on **Disable Sign Ups** if you don't want users to sign up using public endpoints. 
6. Toggle on **Promote Connection to Domain Level** if you want to use this connection with third-party applications.
7. Select **Create**.

For a true passwordless connection, complete the following additional steps:

1. In your new connection, select the **Attributes** tab.
2. To sisable Username as an identifier, select **Configure** and toggle off **Use Username as Identifier**.
3. To configure Email and Phone identifiers, select **Configure**. For Email attributes, enable **Verify email on sign up** is required for true passwordless configuration and ensures users are always prompted for OTP on login and signup.
4. Select **Save**.

> 5. Configure settings for at least **two** authentication methods. You cannot disable password unless you have both Phone and Email configured.

Reviewer 2: We don't need to have two auth methods. We can have a connection with email attribute + email_otp auth method only, which is also a true passwordless connection.

```suggestion
5. Configure settings for authentication methods corresponding to the configured identifiers. You cannot disable password unless you have phone_otp and/or email_otp configured.
```

Ignore if you meant something else here.

Author: No, this is good ^ Added to the next commit. We already mention in this article OTP is the default and addresses SEC-1557
